bip-0352.mediawiki

```diff
@@ -0,0 +1,457 @@
+
+''' P2WPKH '''
+
+The sender performs the tweak using the private key for the output and the receiver obtains the public key from the ''witness''.
+
+''' P2SH-P2WPKH '''
+
+The sender performs the tweak using the private key for the nested ''P2WPKH'' output and the receiver obtains the public key from the ''witness''.
+
+''' P2PKH '''
+
+The sender performs the tweak using the private key for the output and SHOULD sign using the standard script template:
+
+
+    scriptSig: <Signature> <Public Key>
+
+
+The receiver obtains the public key from the ''scriptSig''. The receiver MUST parse the ''scriptSig'' for the public key, even if the ''scriptSig'' is non-standard (e.g. <code><dummy> OP_DROP <Signature> <Public Key></code>). This is to address the [https://en.bitcoin.it/wiki/Transaction_malleability third-party malleability of ''P2PKH'' ''scriptSigs''].
+
+=== Sender ===
+
+==== Selecting inputs ====
+
+The sending wallet performs coin selection as usual with the following restrictions:
+
+* At least one input MUST be from the ''[[#inputs-for-shared-secret-derivation|Inputs For Shared Secret Derivation]]'' list
+* Exclude inputs with witness version > 1 (see ''[[#scanning-transactions|Scanning transactions]]'')
+* For each taproot output spent the sending wallet MUST have access to the private key corresponding to the taproot output key, unless ''H'' is used as the internal public key
+
+==== Creating outputs ====
+
+After the inputs have been selected, the sender can create one or more outputs for one or more silent payment addresses in the following manner:
+
+* Collect the private keys for each input from the ''[[#inputs-for-shared-secret-derivation|Inputs For Shared Secret Derivation]]'' list
+* For each private key ''a<sub>i</sub>'' corresponding to a [https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki BIP341] taproot output, check that the private key produces a point with an even y-value and negate the private key if not<ref name="why_negate_taproot_private_keys">'''Why do taproot private keys need to be checked?''' Recall from [https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki BIP340] that each x-only public key has two corresponding private keys, ''d'' and ''n - d''. To maintain parity between sender and receiver, it is necessary to use the private key corresponding to the even y-value when performing the ECDH step since the receiver will assume the even y-value when summing the taproot x-only public keys.</ref>
+* Let ''a = a<sub>0</sub> + a<sub>1</sub> + … a<sub>n</sub>'', where each ''a<sub>i</sub>'' has been negated if necessary
+* Generate the ''outpoints_hash'', using the method described above
+* Group receiver silent payment addresses by ''B<sub>scan</sub>'' (e.g. each group consists of one ''B<sub>scan</sub>'' and one or more ''B<sub>m</sub>'')
```

Comment:
I think the specification should define the sorting of the `_m_`s in each `_scan_` group. In the reference implementation it just appends the _m_ keys in the order it finds it, meaning that altering the order of the inputs will result in different keys being generated. 

If this is intended it means that the caller is responsible for the order of the arguments, which is asking for trouble I think (I can't see how the receiver could know for sure the order of the labels the sender used). Otoh it's only a problem when sender wants to pay multiple labels of the same receiver in one transaction, which is arguably rare, but still.

See also https://github.com/cygnet3/rust-silentpayments/issues/11

Reply:
You are correct that changing the ordering will change the outputs created, but this isn't a problem for the receiver: the receiver does not need to know the ordering of the outputs to be able to find them.

I can see how this would be annoying from a testing perspective, however, because changing the order of the _B<sub>m</sub>_ values will lead to completely different output sets if multiple labels are being paid by the same sender. I'll add a sorting step, but also a footnote that mentions that sorting is not necessary for the protocol but useful for determinism in testing.

I

Reply:
After rereading carefully the bip, I see how the receiver would not be affected by this. 

Actually sorting the labels as I first thought is indeed impossible, as the sender can't be aware of them. It could sort `m_pubkey`s though, but maybe it doesn't worth it. 

If it's only a problem with testing, I think adding it inside the specification might be more confusing than really useful, it's probably better to add it as a comment in the reference implementation, so that future implementer won't get confused like I did. 

Or maybe only in a sidenote here, basically saying : "while modifying the order of `m`s would result in different keys being generated, that doesn't affect the ability of the recipient to find them since it only looks at the difference between his `B_spend` and the actual key in an output, to check if this difference match a `label`.

Reply:
Added a comment to the reference implementation and mentioned this in the testing section